DS-2107: add commit guidelines documentation (#14)
ongov/ontario-design-system: COMMIT-GUIDELINES.md

# Commit Guidelines

Effective commit messages are crucial for maintaining a clear and traceable project history. They not only help current and future developers understand the evolution of the project but also facilitate efficient collaboration and troubleshooting. This document outlines the guidelines for writing commit messages and versioning our software, ensuring consistency and clarity across the project.

By adhering to the Conventional Commit structure and following semantic versioning principles, we can automate changelog generation, streamline the release process, and maintain a high standard of documentation. These practices are essential for the long-term success and maintainability of our project.

> Happy Committing!

Author: I'm going to leave this one as it sounds more cheerful, like _Happy Birthday!_

## Commit messages

Commit messages are the messages that accompany a commit and help to document the flow of the project history. A good commit message goes a long way to helping other developers (or even future you) trace how a project has progressed over time.

This project follows the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) commit message structure. It is automatically linted using a [Husky](https://typicode.github.io/husky/) pre-commit script that validates the message format, ensuring that it follows the Conventional Commit structure along with a few rules (based off the [@commitlint/config-conventional](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional) rule set):

- subjects start with a lower case character,
- subject lines are not longer than 100 characters,
- body lines are not longer than 100 characters (paragraphs can be made by adding single new lines),

> - there are spaces between the subject and the body and the body and the footer created with new lines,

Reviewer: @kyle-binger-3 Except in French :)

These rules help to keep messages concise and on topic.

Messages can be short and sweet, or more verbose to capture what happened in the commit and why it happened.

It is recommended that commit messages:

- Use the present tense ("add feature" not "added feature").

> - Use the imperative mood ("move cursor to..." not "moves cursor to...").

Author: As much as I don't like this style guide rule I'll remove them

### General tip for writing commit messages

1. **Be Concise**: Keep your commit messages brief and to the point. Avoid unnecessary details.
1. **Focus on the Why**: Explain why the change was made, not just what was changed. This helps others understand the context.
1. **Avoid Vague Language**: Be specific about what the commit does. Avoid terms like "fixed stuff" or "updated code".
1. **Use Consistent Formatting**: Stick to the agreed-upon format for commit messages to maintain consistency.

> 1. **Review Before Committing**: Take a moment to review your commit message before finalizing it. Ensure it accurately reflects the changes made.

Author: Markdown autoformates them as a properly numbered list in output 🙂

## Semantic versioning

Semantic versioning is a way of versioning software that follows a structured version numbering convention. It helps identify the significance between versions and is always forward facing, meaning versions are always moving into the future.

The semantic versioning scheme uses three digits to construct the version number. For example, say we had version `5.2.6`:

```
  5  .  2  .  6
Major.Minor.Patch
```

The first digit is the _Major_ version, which is incremented whenever there are breaking changes.

The second digit is the _Minor_ version, which is incremented when there are backwards compatible changes, eg. a new component.

The third digit is the _Patch_ version, which is incremented when there are backwards compatible bug fixes.

> 

Reviewer: Is this relevant to the subject (commits and guidelines) and useful to the audience?

### Changelog generation

> The commit messages are not only consumed by humans, they are also consumed by the project's tooling when publishing our packages upon release. The project uses [semantic-release](https://github.com/semantic-release/semantic-release) to automatically version the packages, create a changelog for each release, and publish the packages to npm. It does this by tracking the Conventional Commit prefixes on each commit to decide its impact.

Reviewer: ```suggestion
The commit messages are also consumed by the project's tooling when publishing our packages upon release; the Conventional Commit prefixes are used to decide each commit's impact. This is done by [semantic-release](https://github.com/semantic-release/semantic-release) when it automatically versions the packages, creates a changelog for the release, and publishes the relevant packages to npm.
```

> The two relevant prefixes are `fix` and `feat`. Commits with these prefixes should have clean messages about what they are fixing or what feature is being added, respectively. Each commit specifies the weight and impact of the commit on the project.

Reviewer: ```suggestion
The two most frequently used prefixes tend to be `fix` and `feat`. Commits with these prefixes should have clean messages about what they are fixing or what feature is being added, respectively. Each commit specifies the weight and impact of the commit on the project. `docs`, and `refactor` are frequently used, as well. 
```

Commits prefixed with `fix` are considered bug fixes, and so they are deemed to have little impact on the project, are backwards compatible with the previous version, and aren't going to cause any breaking changes. These commits trigger a patch version to be generated, which is version increment of `0.0.1` in the semantic versioning scheme.

Commits prefixed with `feat` are considered features, which introduce new functionality into the project, such as a new component, new style change, or other significant change. These commits are backwards compatible as well and trigger a version increment of `0.1.0`.

> Not every commit needs to start with `fix` or `feat` within a branch, only the ones that have changes worth calling out in our `CHANGELOG.md` file. Commits can be prefixed with other prefixes: `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`. These are more suited for smaller parts of a larger update or commits that are being made that should not impact the version number, example project documentation updates.

Reviewer 2: Need to add `for` before `example`.

e.g. `for example project documentation updates.`

Author: Good catch! 🎣

> Not every commit needs to start with `fix` or `feat` within a branch, only the ones that have changes worth calling out in our `CHANGELOG.md` file. Commits can be prefixed with other prefixes: `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`. These are more suited for smaller parts of a larger update or commits that are being made that should not impact the version number, example project documentation updates.

Author: This is a good question, let me do some research and I'll update the documentation with it.  

Off the top of my head:
- `build` is for build related changes - ie. the building of the code within the project.  These changes are behind the scenes and don't impact the change log, so they get captured but not visibly to the end user.
- `ci` is for CI/CD updates to the workflows/pipelines.  These are specific to only things that that are needed to make the CI/CD system run smoothly.  In this project's case, files in the `.github/workflows` folder.

> Not every commit needs to start with `fix` or `feat` within a branch, only the ones that have changes worth calling out in our `CHANGELOG.md` file. Commits can be prefixed with other prefixes: `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`. These are more suited for smaller parts of a larger update or commits that are being made that should not impact the version number, example project documentation updates.

Reviewer 3: This looks great! Thanks for updating this Scott.

> Not every commit needs to start with `fix` or `feat` within a branch, only the ones that have changes worth calling out in our `CHANGELOG.md` file. Commits can be prefixed with other prefixes: `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, or `test:`. These are more suited for smaller parts of a larger update or commits that are being made that should not impact the version number, for example project documentation updates.

Reviewer: ```
Commits prefixed with `docs` are documentation only changes. Commits prefixed with `refactor` are code changes that neither fix a bug nor add a feature.
```

Author: ```suggestion
Not every commit needs to start with `fix` or `feat` within a branch, only the ones that have changes worth calling out in our `CHANGELOG.md` file. Commits can be prefixed with other prefixes: `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, or `test:`. These are more suited for smaller parts of a larger update or commits that are being made that should not impact the version number, for example project documentation updates. The other prefixes have particular meanings, such as commits prefixed with `docs` are documentation only changes, commits prefixed with `refactor` are code changes that neither fix a bug nor add a feature, and commit prefixed with `ci` are code changes within the CI/CD workflow files.
```

> The changelog can be found within the [CHANGELOG.md](https://github.com/ongov/ontario-design-system/blob/develop/CHANGELOG.md) file, which a running history of all changes made in each release. The changelog version entries look like the following example:

Reviewer 2: Need to add `is` between `which` and `a`.

e.g. `which is a running history of all changes`

```
# 5.0.0-alpha.3 (2024-12-17)

### Bug Fixes

- **grid.mixins:** declare $modules before use 1c7791f
- **utils:** isNumber function validation inverted 37637b6
- **utils:** update `retrieveEnumKeys` function 325d3af

> ```

Reviewer: Might be worth including the example commits that would have been parsed to create this changelog entry.

#### Breaking changes

> A breaking change is a change that is significant enough to cause an incompatibility with the current state of the packages. This could be a removed component property, a set of styles that work differently, or a hole host of different changes.

Reviewer 2: `hole` should be `whole`.

Author: 🕳️

Breaking changes are denoted in the footer of the commit by starting it with `BREAKING CHANGE:` followed by a message about what the breaking change impacts.

> It is highly recommended to add a clear reason for the breaking change and what actions need to be taken when upgrading to the newer version. See the example under [Breaking change](#breaking-change) for an idea of how to communicate the change.

Reviewer: Similar to an earlier comment, instead of recommending we should just set prescriptive guidelines.

```suggestion
Breaking change commits must include a clear reason for the breaking change and what actions need to be taken when upgrading to the newer version. See the example under [Breaking change](#breaking-change) for an idea of how to communicate the change.
```

> It is important to note: Breaking changes can be added to any commit irregardless of the commit prefix, they are not just relegated to `fix` or `feat` commit prefixes as those aren't always relevant when introducing a breaking change.

Reviewer 2: Consider bolding `It is important to note:`

Author: TIL! This is one of the contexts I use _irregardless_ to emphasis the lack of regard that needs to be taken, but I've updated it nonetheless.

> It is important to note: Breaking changes can be added to any commit irregardless of the commit prefix, they are not just relegated to `fix` or `feat` commit prefixes as those aren't always relevant when introducing a breaking change.

Reviewer 2: Great section here. Does it make sense to add the note about adding exclamation marks after the commit prefix as a best practice? Even if our tooling doesn't pick it up or look for it?

Author: Good thought!

> The Conventional Commit structure also allows for an exclamation mark (`!`) to be added before the `:` in the subject. This pattern isn't respected by this projects tooling but it can help commits that contain breaking changes stand out when reviewing the commit subjects.

Reviewer: If we are not using this, should we even mention that this is a thing? Or should we rewrite this simply as:
```suggestion
While the Conventional Commit standards allow for an exclamation mark (`!`) to be added before the `:` in the subject of a commit, this is not a pattern followed by this project.
```

> Good commit messages go along way in creating and maintaining a high quality project. They tell the story of how it grew and evolved overtime. They also help when investigating where things happened, should you ever have to go back in time. Commit messages are almost as important as high quality documentation and code comments.

Reviewer 2: `along` should become `a long` and `overtime` should become `over time`.

> Good commit messages go a long way in creating and maintaining a high quality project. They tell the story of how it grew and evolved over time. They also help when investigating where things happened, should you ever have to go back in time. Commit messages are almost as important as high quality documentation and code comments.

Reviewer: ```suggestion
Good commit messages are important for creating and maintaining a high quality project. They tell the story of how the project grew and evolved over time. They also help when investigating where things happened, should you ever have to go back in time. Commit messages are as important as high quality documentation and code comments.
```

### Base structure

The base structure of a commit message is:

```

> prefix(optional-scope): subject

Reviewer: I am a strong proponent of scope and think we, as a team, could benefit from prescribing this as required. 
Example scopes:
`monorepo`, `app-react`, `ontario-design-system` (as the alternative to `monorepo`), `ontario-design-system-component-library` or `component-library`, etc.

Author: I'm trying to follow the Conventional Commit guidelines, and so I've left leeway here for flexibly as I've found the scope isn't always relevant.  Perhaps this is due to too big of a scope change within in a commit, but I have tried to encourage the use of the scope first, and covered the optional missing scope case later.

The tooling will automatically scope the commits if the scope is missing, it draws this from the packages the commit impacts.

footer
```

### Commits with prefix and scope

Adding a prefix and scope helps indicate the impact and theme of the commit and what it impacts.

Commits can be simple:

```
chore(monorepo): add useWorkspaces to lerna.json
```

Commits can be more detailed:

```
fix(utils): isNumber function validation inverted

This fix handles validating numbers are numbers and not numbers are `NaN`.

- Add `undefined` and `null` checks to `isNumber` to guarantee they're
  false.
- Add unit tests to validate this case via the `date-validation-utils` function `isInvalidYear`.
- Add unit tests to validate the `isNumber` util function.

JIRA: DS-1234
```

Commits can have body text span multiple lines as paragraphs using new lines:

```
fix(utils): update retrieveEnumKeys function

> Updates to `isNumeric` invalided the logic this function was using,

Reviewer 2: Very minor change here but I believe `isNumeric` should instead be `isNumber` as that is what is referenced above.

Author: I think I made a typo in the original commit 😅 I'll update this sample, but I think it is going down in history! [Yup, I did 🙃](https://github.com/ongov/ontario-design-system/commit/325d3affd394474f3ea20fb97c16aa91b22d44e3)!

inverting it restores intended functionality.

- Add unit tests to validate different types of enum cases as well as
  `null` and `undefined`.
```

### Breaking change

Adding `BREAKING CHANGE:` to the footer denotes the commit has a breaking change and that there should be a major version increment.

```
feat(ontario-card)!: added layout prop
 - Updated unit tests
 - Regenerated component definitions

BREAKING CHANGE: `card-type` prop is now deprecated. Use `layout` prop instead.

> ```

Reviewer 2: Similar to my note in the bigger Breaking Change section above, is it worthwhile to put a note about the exclamation mark at the end of a commit prefix as a best practice?

e.g. It is also a best practice to include an exclamation mark after your commit prefix/scope. This is not used as an identifier by the changelog tooling (instead the BREAKING CHANGE footer identifier is leveraged), but it can be used as a further identifier to the reader that the change is of a breaking nature.

Author: I seem to have forgotten to do this, I'll add something based on your suggestion.

BREAKING CHANGE: `card-type` prop is now deprecated. Use `layout` prop instead.
```

> 

Author: ```suggestion

_Note_: It is a best practice to include an exclamation mark (`!`) after your commit `type(scope)`. This is not used as an identifier by the changelog tooling (instead the `BREAKING CHANGE` footer identifier is leveraged), but it can be used as a further identifier to the reader that the change is of a breaking nature when reviewing the subjects.

```

> Sometimes commits span more than a single scope but contain a number of changes within the same theme, or action, or the scope is ambiguous. When this happens it is recommended to break the commits down by scope, but in the case the commits impact multiple scopes the same commit message rules apply but the scope can be emitted.

Reviewer 2: This is a good section but I think it can be written to flow a bit better. Let me knoww hat you think of my try at it below:

```
Sometimes commits can span more than a single scope but contain a number of changes within the same theme or action, or the scope is ambiguous. When this happens, it is recommended to break the commits down by scope. However, in cases where the commits impact multiple scopes, the scope can be omitted.  All other commit message rules still apply.
```

Also `emitted` should be `omitted`.

> Sometimes commits can span more than a single scope but contain a number of changes within the same theme or action, or the scope is ambiguous. When this happens, it is recommended to break the commits down by scope. However, in cases where the commits impact multiple scopes, the scope can be omitted. All other commit message rules still apply.

Reviewer: ```suggestion
Commits will occasionally span more than a single scope but contain a number of changes within the same theme or action, or the scope is ambiguous. When this happens, it is recommended to break the commits down by scope. However, in cases where the commits impact multiple scopes, the scope can be omitted. All other commit message rules still apply.
```

Reviewer: `When this happens, it is recommended to break the commits down by scope. However, in cases where the commits impact multiple scopes, the scope can be omitted. All other commit message rules still apply.`

I think we should require commits being broken down by scope to keep commits within a proper scope of impact. Commits that impact multiple scopes should be broken down, if they cannot then they are a monorepo scoped change.

Author: This is an edge case, and generally we shouldn't end up here too often.  The one that comes to mind is the `ci` type, as we don't have a very big scope for the CI so it is pretty clear what it is impacting.  Perhaps we could do `ci(main.yml) or something like this, but usually changes there happen across the entire CI file set and the `scope` doesn't add value.

I'm curious your thoughts on this edge case and how we might be able to tackle it?

- body lines must not be longer than 100 characters (paragraphs can be made by adding single new lines)
- new lines must exist between the subject and the body, as well as the body and the footer

These rules keep messages concise and relevant.

Messages can be short and sweet, or more verbose to capture what happened in the commit and why it happened.

Commit messages should:

- use the present tense ("add feature" not "added feature")
- use the imperative mood ("move cursor to..." not "moves cursor to...")

### General tips for writing commit messages

1. **Be Concise**: Keep your commit messages brief and to the point. Avoid unnecessary details.
2. **Focus on the Why**: Explain why the change was made, not just what was changed. This helps others understand the context.
3. **Avoid Vague Language**: Be specific about what the commit does. Avoid terms like "fixed stuff" or "updated code".
4. **Use Consistent Formatting**: Stick to the agreed-upon format for commit messages to maintain consistency.
5. **Review Before Committing**: Take a moment to review your commit message before finalizing it. Ensure it accurately reflects the changes made.

> Do not be afraid to breakup files across multiple commits they they thematically do not fall into one unit.

Reviewer 4: ```suggestion
Do not be afraid to break up files across multiple commits if they thematically do not fall into one unit.
```

## Semantic versioning

> Semantic versioning is a way of versioning software that follows a structured version numbering convention. It helps identify the significance between versions and is always forward facing, meaning versions are always moving into the future.

Reviewer 4: ```suggestion
Semantic versioning is a way of versioning software that follows a structured version numbering convention.It helps determine the significance of changes between versions and ensures that version numbers always increase over time.
```

```
  5  .  2  .  6
Major.Minor.Patch
```

The first digit is the _Major_ version, which is incremented whenever there are breaking changes.

The second digit is the _Minor_ version, which is incremented when there are backwards compatible changes, eg. a new component.

The third digit is the _Patch_ version, which is incremented when there are backwards compatible bug fixes.

### Changelog generation

The commit messages are not only consumed by humans, they are also consumed by the project's tooling when publishing our packages upon release. The project uses [semantic-release](https://github.com/semantic-release/semantic-release) to automatically version the packages, create a changelog for each release, and publish the packages to npm by tracking the Conventional Commit prefixes on each commit to decide its impact.

The two relevant prefixes are `fix` and `feat`. Commits with these prefixes should have clean messages about what they are fixing or what feature is being added, respectively. Each commit specifies the weight and impact of the commit on the project.

> Commits prefixed with `fix` are considered bug fixes, and so they are deemed to have little impact on the project, are backwards compatible with the previous version, and aren't going to cause any breaking changes. These commits trigger a patch version to be generated, which is a version increment of `0.0.1` in the semantic versioning scheme.

Reviewer 4: ```suggestion
Commits prefixed with `fix` are considered bug fixes. They are deemed to have little impact on the project, are backward-compatible with the previous version, and do not cause any breaking changes. These commits trigger a patch version increment of 0.0.1 in the semantic versioning scheme.
```

> Commits prefixed with `feat` are considered features, which introduce new functionality into the project, such as a new component, new style change, or other significant change. These commits are backwards compatible as well and trigger a version increment of `0.1.0`.

Reviewer 4: ```suggestion
Commits prefixed with `feat` are considered features, which introduce new functionality into the project, such as a new component, new style change, or other significant change. These commits are backwards-compatible as well and trigger a version increment of `0.1.0`.
```

Author: I've never seen this written with a hyphen before 🤔  They are generally two words

> Not every commit needs to start with `fix` or `feat` within a branch, only the ones that have changes worth calling out in our `CHANGELOG.md` file. See [Non `feat` or `fix` commit prefixes](#non-feat-or-fix-commit-prefixes) for more information.

Reviewer 4: May be worth adding an example of this below here.

Author: I'll add an example in the _Examples_ section of a non `feat`/`fix` commit, thanks for the idea! 💡

The changelog can be found within the [CHANGELOG.md](https://github.com/ongov/ontario-design-system/blob/develop/CHANGELOG.md) file, which is a running history of all changes made in each release. The changelog version entries look like the following example:

```
# 5.0.0-alpha.3 (2024-12-17)

### Bug Fixes

- **grid.mixins:** declare $modules before use 1c7791f
- **utils:** isNumber function validation inverted 37637b6
- **utils:** update `retrieveEnumKeys` function 325d3af
```

This example was made from three different commit message subjects:

- (1c7791f) fix(grid.mixins): declare $modules before use
- (37637b6) fix(utils): isNumber function validation inverted
- (325d3af) fix(utils): update `retrieveEnumKeys` function

The `type` is parsed into a section within the release version, the `(scope)` indicates what was impacted, and the `subject` describes the specific change within that scope.

#### Breaking changes

> A breaking change is a change that is significant enough to cause an incompatibility with the current state of the packages. This could be a removed component property, a set of styles that work differently, or a whole host of different changes.

Reviewer 4: \
```suggestion
A breaking change is a change significant enough to cause an incompatibility with the current state of the package(s). This could be the removal of a component property, a modification to existing styles, or various other impactful changes.
```
